### PR TITLE
Restructure and strengthen guidance on conditional requests

### PR DESCRIPTION
**Pull request by mikekistler**
This PR restructures and updates the guidance for processing of conditional requests.

The primary motivation for this update is the research that I did on the conditional headers for #458. This led me to realize that we needed to strengthen our guidance so that services would correctly handle these headers whenever they are present.

I also moved much of the explanatory content to a new section in ConsiderationsForServiceDesign, so that the section in the Guidelines could be kept to just the direct and actionable guidance.

### CODE REVIEW

File: azure/ConsiderationsForServiceDesign.md [hunk 1]
```diff
@@ -172,7 +172,7 @@ Before releasing your API plan to invest significant design effort, get customer
...
 
-However, given how many organizations are structured, it's common that this email address is different from the actual people writing code against your API. To address this, the service API should declare that it may return the `azure-deprecating` header, to indicate that this operation will be removed in the future. There is a simple string convention, specifed in the [Azure REST API Guidelines](https://aka.ms/azapi/guidelines) that provides more information about the forthcoming deprecation.
+However, given how many organizations are structured, it's common that this email address is different from the actual people writing code against your API. To address this, the service API should declare that it may return the `azure-deprecating` header, to indicate that this operation will be removed in the future. There is a simple string convention, specified in the [Azure REST API Guidelines](https://aka.ms/azapi/guidelines) that provides more information about the forthcoming deprecation.
```
Comment (mikekistler) [inline azure/ConsiderationsForServiceDesign.md]: Unrelated typo fix.

File: azure/ConsiderationsForServiceDesign.md [hunk 2]
```diff
@@ -234,7 +234,7 @@ PATCH must never be used for long-running operations -- it should be reserved fo
 If a long-running update is required it should be implemented with POST.
 
 There is a special form of long-running operation initiated with PUT that is described
-in [Create (PUT) with additional long-running processing](#create-put-with-additional-long-running-processing).
+in [Create (PUT) with additional long-running processing](./Guidelines.md#put-operation-with-additional-long-running-processing).
```
Comment (mikekistler) [inline azure/ConsiderationsForServiceDesign.md:238]: Unrelated fix to link.

File: azure/Guidelines.md [hunk 3]
```diff
@@ -799,7 +799,7 @@ While removing a value from an enum is a breaking change, adding value to an enu
...
 ### Deprecating Behavior Notification
 
-When the [API Versioning](#API-Versioning) guidance above cannot be followed and the [Azure Breaking Change Reviewers](mailto:azbreakchangereview@microsoft.com) approve a [breaking change](#123-definition-of-a-breaking-change) to a specific API version it must be communicated to its callers. The API version that is being deprecated must add the `azure-deprecating` response header with a semicolon-delimited string notifying the caller what is being deprecated, when it will no longer function, and a URL linking to more information such as what new operation they should use instead.
+When the [API Versioning](#api-versioning) guidance above cannot be followed and the [Azure Breaking Change Reviewers](mailto:azbreakchangereview@microsoft.com) approve a breaking change to a specific API version it must be communicated to its callers. The API version that is being deprecated must add the `azure-deprecating` response header with a semicolon-delimited string notifying the caller what is being deprecated, when it will no longer function, and a URL linking to more information such as what new operation they should use instead.
```
Comment (mikekistler) [inline azure/Guidelines.md:802]: Unrelated link fix.

File: azure/ConsiderationsForServiceDesign.md [hunk 4]
```diff
@@ -492,6 +492,33 @@ and the number of results to return, respectively.
...
+An Etag identifies a 'version' or 'instance' of a resource and is computed by the service and returned in an `ETag` response header for GET or other operations on the resource.
+
+### Cache Control
+
+One of the more common uses for conditional requests is cache control. This is especially useful when resources are large in size, expensive to compute/calculate, or hard to reach (significant network latency).
+A client can make a "conditional GET request" for the resource, with a precondition header that requests that
+data be returned only when the version on the service does not match the Etag or last modified date in the header.
+If there are no changes, then there is no need to return the resource, as the client already has the most recent version.
+
+Implementing this strategy is relatively straightforward. First, you will return an `ETag` with a value that uniquely identifies the instance (or version) of the resource. The [Computing ETags](./Guidelines.md#computing-etags) section provides guidance on how to properly calculate the value of your `ETag`.
```
Comment (weidongxu-microsoft) [inline azure/ConsiderationsForServiceDesign.md:511]: nit, `[Computing ETags](./Guidelines.md#computing-etags)` already changed to `ETags` in this PR.

File: azure/ConsiderationsForServiceDesign.md [hunk 5]
```diff
@@ -492,6 +492,33 @@ and the number of results to return, respectively.
...
+This tells the service to compare the `ETag` that came in with the request, with the latest value that it has calculated. If the two values are the same, then it is not necessary to return the resource to the client--it already has it. If they are different, then the service will return the latest version of the resource, along with the updated `ETag` value in the header.
+
+### Optimistic Concurrency
+
+Optimistic concurrency is a strategy used in HTTP to avoid the "lost update" problem that can occur when multiple clients attempt to update a resource simultaneously.
+Clients can use ETags returned by the service to specify a _precondition_ for the execution of an update, to ensure that the resource has not been updated since the client last observed it.
+For example, the client can specify an `If-Match` header with the last ETag value received by the client in an update request.
+The service processes the update only if the Etag value in the header matches the ETag of the current resource on the server.
+By computing and returning ETags for your resources, you enable clients to avoid using a "pessimistic" strategy where the "last write always wins."
```
Comment (weidongxu-microsoft) [inline azure/ConsiderationsForServiceDesign.md]: This line may not be accurate. The usual "pessimistic" strategy would be "locking", which likely won't available to client.

Maybe mention about avoiding "last write wins" be enough.

File: azure/Guidelines.md [hunk 6]
```diff
@@ -1020,28 +1020,38 @@ Depending on the requirements of the service, there can be any number of "input"
...
-An `ETag` is always set by the service and will enable you to _conditionally_ control how your service responds to requests, enabling you to provide predictable updates and more efficient access.
 
-<a href="#condreq-return-etags" name="condreq-return-etags">:ballot_box_with_check:</a> **YOU SHOULD** return an `ETag` with any operation returning the resource or part of a resource or any update of the resource (whether the resource is returned or not).
+The [HTTP Standard][] defines request headers that clients may use to specify a _precondition_
+for execution of an operation. These headers allow clients to implement efficient caching mechanisms
+and avoid data loss in the event of concurrent updates to a resource. The headers that specify conditional execution are "If-Match", "If-None-Match", "If-Modified-Since", "If-Unmodified-Since", and "If-Range".
```
Comment (weidongxu-microsoft) [inline azure/Guidelines.md]: nit, wondering whether we need to mention `If-Range` at current stage (I assume it couples with `Range` request/header).

File: azure/ConsiderationsForServiceDesign.md [hunk 7]
```diff
@@ -172,7 +173,7 @@ Before releasing your API plan to invest significant design effort, get customer
...
 
-However, given how many organizations are structured, it's common that this email address is different from the actual people writing code against your API. To address this, the service API should declare that it may return the `azure-deprecating` header, to indicate that this operation will be removed in the future. There is a simple string convention, specifed in the [Azure REST API Guidelines](https://aka.ms/azapi/guidelines) that provides more information about the forthcoming deprecation.
+However, given how many organizations are structured, it's common that this email address is different from the actual people writing code against your API. To address this, the service API should declare that it may return the `azure-deprecating` header, to indicate that this operation will be removed in the future. There is a simple string convention, specified in the [Azure REST API Guidelines](https://aka.ms/azapi/guidelines) that provides more information about the forthcoming deprecation.
```
Comment (heaths) [inline azure/ConsiderationsForServiceDesign.md]: We should give them a bookmark i.e., <https://aka.ms/azapi/guidelines#deprecating-behavior-notification>.

File: azure/ConsiderationsForServiceDesign.md [hunk 8]
```diff
@@ -234,7 +235,7 @@ PATCH must never be used for long-running operations -- it should be reserved fo
 If a long-running update is required it should be implemented with POST.
 
 There is a special form of long-running operation initiated with PUT that is described
-in [Create (PUT) with additional long-running processing](#create-put-with-additional-long-running-processing).
+in [Create (PUT) with additional long-running processing](./Guidelines.md#put-operation-with-additional-long-running-processing).
```
Comment (heaths) [inline azure/ConsiderationsForServiceDesign.md:238]: In our SDK markdowns, we're not allowed to use relative references because they don't work in all contexts. Recommend a bookmarked aka link i.e., <https://aka.ms/azapi/guidelines#put-operation-with-additional-long-running-processing>.

File: azure/ConsiderationsForServiceDesign.md [hunk 9]
```diff
@@ -492,6 +493,33 @@ and the number of results to return, respectively.
 
 Note that when `top` specifies a value larger than the server-driven paging page size, the response will be paged accordingly.
 
+## Conditional Requests
+
+When designing an API, you will almost certainly have to manage how your resource is updated. For example, if your resource is a bank account, you will want to ensure that one transaction--say depositing money--does not overwrite a previous transaction.
+Similarly, it could be very expensive to send a resource to a client. This could be because of its size, network conditions, or a myriad of other reasons.
+Both of these scenarios can be accomplished with conditional requests, where the client specifies a _precondition_
+for execution of a request, based on its last modification date or entity tag ("ETag").
+An Etag identifies a 'version' or 'instance' of a resource and is computed by the service and returned in an `ETag` response header for GET or other operations on the resource.
```
Comment (heaths) [inline azure/ConsiderationsForServiceDesign.md]: Nit: I'd stick with a consistent casing of "ETag" ("ETag" vs. "Etag").

File: azure/ConsiderationsForServiceDesign.md (hunk 8 above)
Comment (heaths) [reply azure/ConsiderationsForServiceDesign.md:238]:
```suggestion
in [Create (PUT) with additional long-running processing](https://aka.ms/azapi/guidelines#put-operation-with-additional-long-running-processing).
```

File: azure/ConsiderationsForServiceDesign.md (hunk 7 above)
Comment (heaths) [reply azure/ConsiderationsForServiceDesign.md]:
```suggestion
However, given how many organizations are structured, it's common that this email address is different from the actual people writing code against your API. To address this, the service API should declare that it may return the `azure-deprecating` header, to indicate that this operation will be removed in the future. There is a simple string convention, specified in the [Azure REST API Guidelines](https://aka.ms/azapi/guidelines#deprecating-behavior-notification) that provides more information about the forthcoming deprecation.
```

File: azure/ConsiderationsForServiceDesign.md [hunk 10]
```diff
@@ -492,6 +493,33 @@ and the number of results to return, respectively.
...
+
+### Cache Control
+
+One of the more common uses for conditional requests is cache control. This is especially useful when resources are large in size, expensive to compute/calculate, or hard to reach (significant network latency).
+A client can make a "conditional GET request" for the resource, with a precondition header that requests that
+data be returned only when the version on the service does not match the Etag or last modified date in the header.
+If there are no changes, then there is no need to return the resource, as the client already has the most recent version.
+
+Implementing this strategy is relatively straightforward. First, you will return an `ETag` with a value that uniquely identifies the instance (or version) of the resource. The [Computing ETags](./Guidelines.md#computing-etags) section provides guidance on how to properly calculate the value of your `ETag`.
```
Comment (heaths) [inline azure/ConsiderationsForServiceDesign.md:511]:
```suggestion
Implementing this strategy is relatively straightforward. First, you will return an `ETag` with a value that uniquely identifies the instance (or version) of the resource. The [Computing ETags](https://aka.ms/azapi/guidelines#computing-etags) section provides guidance on how to properly calculate the value of your `ETag`.
```

File: azure/Guidelines.md (hunk 3 above)
Comment (heaths) [inline azure/Guidelines.md:802]:
```suggestion
When the [API Versioning](https://aka.ms/azapi/guidelines#api-versioning) guidance above cannot be followed and the [Azure Breaking Change Reviewers](mailto:azbreakchangereview@microsoft.com) approve a breaking change to a specific API version it must be communicated to its callers. The API version that is being deprecated must add the `azure-deprecating` response header with a semicolon-delimited string notifying the caller what is being deprecated, when it will no longer function, and a URL linking to more information such as what new operation they should use instead.
```

File: azure/Guidelines.md [hunk 11]
```diff
@@ -1020,28 +1020,40 @@ Depending on the requirements of the service, there can be any number of "input"
...
 ### Conditional Requests
-When designing an API, you will almost certainly have to manage how your resource is updated. For example, if your resource is a bank account, you will want to ensure that one transaction--say depositing money--does not overwrite a previous transaction.
-Similarly, it could be very expensive to send a resource to a client. This could be because of its size, network conditions, or a myriad of other reasons. To enable this level of control, services should leverage an `ETag` header, or "entity tag," which will identify the 'version' or 'instance' of the resource a particular client is working with.
-An `ETag` is always set by the service and will enable you to _conditionally_ control how your service responds to requests, enabling you to provide predictable updates and more efficient access.
 
-<a href="#condreq-return-etags" name="condreq-return-etags">:ballot_box_with_check:</a> **YOU SHOULD** return an `ETag` with any operation returning the resource or part of a resource or any update of the resource (whether the resource is returned or not).
+The [HTTP Standard][] defines request headers that clients may use to specify a _precondition_
```
Comment (heaths) [inline azure/Guidelines.md:1024]: Tip: at least with GitHub-flavored markdown, if you use an anchor like this you don't need the empty brackets i.e.,
```suggestion
The [HTTP Standard] defines request headers that clients may use to specify a _precondition_
```

File: azure/Guidelines.md [hunk 12]
```diff
@@ -1020,28 +1020,40 @@ Depending on the requirements of the service, there can be any number of "input"
...
+for execution of an operation. These headers allow clients to implement efficient caching mechanisms
+and avoid data loss in the event of concurrent updates to a resource. The headers that specify conditional execution are "If-Match", "If-None-Match", "If-Modified-Since", "If-Unmodified-Since", and "If-Range".
```
Comment (heaths) [inline azure/Guidelines.md]: Nit: since these are programmatic constructs like `ETag`, I'd use backticks i.e.,
```suggestion
and avoid data loss in the event of concurrent updates to a resource. The headers that specify conditional execution are `If-Match`, `If-None-Match`, `If-Modified-Since`, `If-Unmodified-Since`, and `If-Range`.
```
You have been for other programmatic constructs so far e.g., `azure-deprecating`.

File: azure/Guidelines.md (hunk 3 above)
Comment (mikekistler) [reply azure/Guidelines.md:802]: I'm going to hold off on changing relative to absolute references until it's clear that this is necessary. If/when we do it we should make the change throughout.

**Review by JeffreyRichter — APPROVED** [review]

**Review by markweitzel — APPROVED** [review]
LGTM. Thank you for pulling this together.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by heaths** [comment]
@markweitzel can you merge this? Board agrees it's ready.
